src/sctools/test/test_bam.py

```diff
@@ -157,3 +157,41 @@ def test_split_succeeds_with_raise_missing_false_and_no_cr_barcode_passed(tagged
     os.remove(tagged_bam)  # clean up
     for f in glob.glob('test_output_*'):
         os.remove(f)
+
+
+def test_tags_queryname_sort_order_gives_correct_order():
+    _, records = bam.read_bam(data_dir + 'queryname-sorted.bam')
+    sorted_records = bam.sort_by_tags_and_queryname(records, ['UB', 'CB', 'GE'])
+
+    # Verify sort order by asserting that tags and queryname are in non-decreasing blocks.
+    old_ub = ''
+    old_cb = ''
+    old_ge = ''
+    old_qn = ''
+    for record in sorted_records:
+        ub = bam.get_tag_or_default(record, 'UB')
+        cb = bam.get_tag_or_default(record, 'CB')
+        ge = bam.get_tag_or_default(record, 'GE')
+        qn = record.query_name
+
+        assert ub >= old_ub
+        if ub > old_ub:
+            cb = ''
```

Comment:
Can you clarify why these tags get set to `''`? If the next record has the same `UB` but the `CB` is not in the right order would this throw off the comparison?

Reply:
If ub has changed, it means cb can be anything and still be in the correct order. We set cb/old_cb to `''` so that for the next line any value of cb will be accepted. Will add some commenting to the code.

```
ub: 1 cb: 1
ub: 1 cb: 2
ub: 2 cb: 1
ub: 2 cb: 2
ub: 3 cb: 1
```

Reply:
Ok, that part makes sense! I was thinking of how to check for the situation where the ub doesn't change and the cb is not in the right order:
```
ub: 1 cb: 1
ub: 1 cb: 2
ub: 2 cb: 2  
ub: 2 cb: 1 <- cb sort order is incorrect
ub: 3 cb: 1
```

Reply:
Yeah, so that's why we check that ub has changed in that first conditional on 178 and then only set cb to `''` if ub has changed.

Reply:
Thinking about this some more, you are totally right! There is a bug here. Going to revamp this.